apps/wing/src/commands/pack.ts

```diff
@@ -0,0 +1,77 @@
+import * as path from "path";
+import * as fs from "fs";
+import * as cp from "child_process";
+
+export interface PackageOptions {
+  /**
+   * Directory to to save the generated package to.
+   */
+  readonly outdir: string;
+}
+
+export async function pack(options: PackageOptions): Promise<string> {
+  // check npm is installed
+  try {
+    cp.execSync("npm --version", { stdio: "ignore" });
+  } catch {
+    throw new Error(`npm is not installed. Install it from https://www.npmjs.com/get-npm`);
+  }
+
+  // check there is at least one source file
+  const sourceFiles = fs.readdirSync(process.cwd()).filter((f) => f.endsWith(".w"));
+  if (sourceFiles.length === 0) {
+    throw new Error(`No Wing source files (.w) found in the current directory.`);
+  }
+
+  // check package.json exists
+  const currentDir = process.cwd();
+  const pkgJsonPath = path.join(currentDir, "package.json");
+  if (!fs.existsSync(pkgJsonPath)) {
+    throw new Error(`No package.json found in the current directory. Run \`npm init\` first.`);
+  }
+
+  const pkgJson = JSON.parse(fs.readFileSync(pkgJsonPath, "utf8"));
+  pkgJson.wing = pkgJson.wing ?? true;
+
+  // check source files will be included in tarball
+  const pkgJsonFiles = pkgJson.files ?? [];
+  const expectedGlobs = ["**/*.js", "**/*.w", "!/target/**"];
+  for (const glob of expectedGlobs) {
+    if (!pkgJsonFiles.includes(glob)) {
+      pkgJsonFiles.push(glob);
+    }
+  }
+  pkgJson.files = pkgJsonFiles;
+
+  // check package.json has required fields
+  const requiredFields = ["name", "version", "description", "author", "license"];
+  for (const field of requiredFields) {
+    if (pkgJson[field] === undefined) {
+      throw new Error(`Missing required field "${field}" in package.json`);
+    }
+  }
+
+  // check if "main" points to a valid file, and if not, create a dummy file
+  let main = pkgJson.main ?? "index.js";
+  if (!fs.existsSync(main)) {
+    const lines = [];
+    lines.push("// This file was generated by wing pack");
+    lines.push(
+      `console.log("${pkgJson.name} is a winglang library and cannot be used from javascript (yet)")`
+    );
+    lines.push();
+    fs.writeFileSync(main, lines.join("\n"));
```

Comment:
Not sure I like the idea of writing index.js and package.json into the project directory.

Perhaps we should create a temporary staging directory? This will give us more control over exactly what goes into the package.

(With the price of having to be more explicit about the anatomy of a wing npm package, which is good in my opinion)

Reply:
I feel like there are two high level approaches we can take for the MVP:

1. Have a dedicated Wing configuration file, like `wing.toml`. This is where you can specify the library name, author, dependencies, license, description, etc. There's no `package.json` or `index.js` file in the project directory. Clean. 
2. Use the existing JS ecosystem - every Wing project has a `package.json` file. We're using the npm ecosystem to manage dependencies, so folks can control this file and reuse `package.json` snippets from their previous JS experiences, like adding custom scripts etc (this could be a good DX for folks coming from JS/TS, worse DX for folks not coming from JS/TS). Running `wing pack` performs a few validations on `package.json`, telling you if anything is configured wrong -- and it might "own" one or two fields, like `version`, but besides that, what the user wrote in `package.json` is what goes in the tarball. No temporary staging directory.

Do you have an opinion on what direction we should follow? (If not I'm happy to continue down option (2) since I wasn't sure if you had warmed up to (1))

Reply:
I'm not sure I see the value in a hybrid approach where there's a package.json in the project directory, but some subset of the fields are copied into a new `package.json` inside of a staging directory. It seems a little surprising (users have to look into docs if they're trying to figure out why a field wasn't included in their tarball). I'm also reminded of how users couldn't customize `tsconfig.json` in JSII, which was a large pain point personally :-/

Reply:
I am not suggesting to create a new package.json and only copy certain fields, just to clone the project directory into a staging area and make modifications there instead of the user code.

It will give us more control over the package structure without polluting the user project, but the DX is will going to be pretty standard package.json.

I am less comfortable with modifying the user directory because it takes us into Projen-land and eventually will require things like self mutations, read only files, and maybe we should try to avoid this (for now).

Once we would want to add jsii manifests, api docs and JavaScript files, there will be more stuff we will want to include in the tarball that we shouldn't just spit into the user directory.

Stuff like this...

If we do `wing.toml` (which we will likely add at some point), we will also have to pack from a staging directory anyway, so staging is basically a step into the right direction regardless of the manifest format I think.

Reply:
> just to clone the project directory into a staging area and make modifications there instead of the user code.

> If we do wing.toml (which we will likely add at some point), we will also have to pack from a staging directory anyway, so staging is basically a step into the right direction regardless of the manifest format I think.

I guess I just wonder if there are any ways we can avoid cloning the user's entire directory. I worry it may be pretty heavy handed as a solution. Sure, we can use excluding heuristics like "ignore node modules" and "ignore .git", but in general there can always be extra stuff in the user's current directory (files for CI, editor configs, helper scripts, symlinks, etc.).

The safer approach might be to just copy files that our compiler is actually seeing as part of the compilation (all of the .w and referenced .js files, plus package.json for now), but I guess we can refine upon this later.